File: examples/json2mss.js
const fs = require('fs');

function createQualifier(qualifierKey, value) {
    const ret = [];
    if (Array.isArray(value)) { // For array data
        for (const v of value) {
            ret.push(["", "", "", qualifierKey, v]);
        }
    } else { // For string
        ret.push(["", "", "", qualifierKey, value]);
    }
    return ret;
}

function createFeature(featureName, featureValues) {
    let ret = [];
    if (Array.isArray(featureValues)) { // For array data (REFERENCE and COMMENT)
        for (const v of featureValues) {
            ret = ret.concat(createFeature(featureName, v));
        }
    } else {
        for (const qualifierKey in featureValues) {
            if (featureValues.hasOwnProperty(qualifierKey)) {
                const value = featureValues[qualifierKey];
                ret = ret.concat(createQualifier(qualifierKey, value));
            }
        }
    }
    if (ret.length > 0) {
        ret[0][1] = featureName;
    }
    return ret;
}

function createCommon(commonJson) {
    let ret = [];
    for (const featureName in commonJson) {
        if (commonJson.hasOwnProperty(featureName) && !featureName.startsWith("_")) {
            const featureValues = commonJson[featureName];
            ret = ret.concat(createFeature(featureName, featureValues));
        }
    }
    if (ret.length > 0) {
        ret[0][0] = "COMMON";
    }
    return ret;
}

// main
const args = process.argv.slice(2);
if (args.length === 0) {
    console.error("Please provide the path to the JSON file as an argument.");
    process.exit(1);
}

const inputJson = args[0];

fs.readFile(inputJson, 'utf8', (err, data) => {
    if (err) {
        console.error("Error reading file:", err);
        return;
    }
    const commonJson = JSON.parse(data);
    const common = createCommon(commonJson);

    common.forEach(row => {
        console.log(row.map(String).join("\t"));
    });
});

File: examples/json2mss.py
import json

def create_qualifier(qualifier_key: str, value: str|list) -> list[list[str]]:
    ret = []
    if isinstance(value, list):  # For array data
        for v in value:
            ret.append(["", "", "", qualifier_key, v])
    else:  # For string
        ret.append(["", "", "", qualifier_key, value])
    return ret

def create_feature(feature_name: str, feature_values: dict|list) -> list[list[str]]:
    ret = []
    if isinstance(feature_values, list):  # For array data (REFERENCE and COMMENT)
        for v in feature_values:
            ret.extend(create_feature(feature_name, v))
    else:
        for qualifier_key, value in feature_values.items():
            ret.extend(create_qualifier(qualifier_key, value))
    if ret:
        ret[0][1] = feature_name
    return ret

def create_common(common_json: dict) -> list[list[str]]:
    ret = []
    for feature_name, feature_values in common_json.items():
        if feature_name != "trad_submission_category":
            ret.extend(create_feature(feature_name, feature_values))
    if ret:
        ret[0][0] = "COMMON"
    return ret

if __name__ == "__main__":
    import sys
    input_json = sys.argv[1]

    with open(input_json, "r") as f:
        common_json = json.load(open(input_json, "r"))
    common = create_common(common_json)

    for row in common:
        row = map(str, row)
        print("\t".join(row))

